core/Enclave/enclave_hsm.cpp

```diff
@@ -458,27 +554,103 @@ sgx_status_t enclave_rsa_encrypt(const uint8_t *cmk_blob,
                                  uint8_t keyspec,
                                  uint8_t padding_mode)
 {
-    const rsa_key_data_t *temp_pub_key = (rsa_key_data_t *)cmk_blob;
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+    
+    // check padding_mode and keyspec
+    //
+    if (keyspec == NULL || padding_mode == NULL) {
+        printf("ecall rsa_encrypt keyspec or padding_mode is wrong.\n");
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
 
-    BIO *bio = BIO_new_mem_buf(temp_pub_key->rsa_public_key, -1);
+    // check plaintext and plaintext_len
+    //
+    if (plaintext == NULL || plaintext_len == 0) {
+        printf("ecall rsa_encrypt plaintext or len is wrong.\n");
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
 
-    RSA* rsa = PEM_read_bio_RSAPublicKey(bio, &rsa, NULL, g_domain_key);
+    uint32_t max_plaintext_len = ehsm_get_rsa_max_encryption_size(keyspec, padding_mode);
+    if (max_plaintext_len == 0)
+        return SGX_ERROR_INVALID_PARAMETER;
+    if (plaintext_len > max_plaintext_len) {
+        printf("ecall rsa_encrypt plaintext_len is up to %d.\n", max_plaintext_len);
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
 
-    if(rsa == NULL)
-    {
-        printf("failed to load rsa key\n");
+    // check ciphertext and ciphertext_len
+    //
+    if (ciphertext == NULL || ciphertext_len < ehsm_get_rsa_cipher_len(keyspec)) {
+        printf("ecall rsa_encrypt ciphertext len is too small, it should be %d\n", ehsm_get_rsa_cipher_len(keyspec));
         return SGX_ERROR_INVALID_PARAMETER;
     }
+    
+    // check cmk_blob and cmk_blob_size
+    //
+    uint32_t encrypted_rsa_len = ehsm_calc_gcm_data_size(0, ehsm_get_rsa_key_pem_size(keyspec)); // ehsm_get_rsa_key_pem_size() will return INT_MAX if keyspec is invalid
+    
+    if (UINT32_MAX == encrypted_rsa_len) {
+        printf("ecall rsa_encrypt failed to calculate encrypted data size.\n");
+        return SGX_ERROR_UNEXPECTED;
+    }
 
-    switch (keyspec)
-    {
-        case RSA_NO_PADDING:
-            RSA_public_encrypt(plaintext_len, plaintext, ciphertext, rsa, RSA_NO_PADDING);
-        case RSA_PKCS1_OAEP_PADDING:
-            RSA_public_encrypt(plaintext_len, plaintext, ciphertext, rsa, RSA_PKCS1_OAEP_PADDING);
+    if (cmk_blob == NULL || cmk_blob_size < encrypted_rsa_len) {
+        printf("ecall rsa_encrypt cmk_blob_size is too small.\n");
+        return SGX_ERROR_INVALID_PARAMETER;
     }
 
-    return SGX_SUCCESS;
+    // do rsa public key encrypt
+    //
+    uint8_t* enc_key = NULL;
+    uint8_t* rsa_public_key = NULL;
+    BIO *bio = NULL;
+    RSA* rsa_pubkey = NULL;
+    uint32_t public_key_size;
+    
+    do {
+        // load rsa public key
+        //
+        public_key_size = ehsm_get_rsa_public_key_pem_size(keyspec);
+        
+        enc_key = (uint8_t*)malloc(cmk_blob_size);
```

Comment:
enc_key is the plaintext to be decrypted.
cmk_blob_size is the size of cmk_blob. The two have nothing relationship.

Reply:
you can refer to https://github.com/intel/ehsm/blob/c665aa19f89988080493a6c82913d7b2af73e9d8/core/Enclave/enclave_hsm.cpp#L142
For GCM, the ciphertext len is same to plaintext len. you can use it to get the plaintext len.

Reply:
copy, i will check this